lib/spack/spack/report.py

```diff
@@ -0,0 +1,283 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/spack/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+"""Tools to produce reports of spec installations"""
+import collections
+import functools
+import itertools
+import os.path
+import time
+
+import llnl.util.lang
+import spack.build_environment
+import spack.fetch_strategy
+import spack.package
+
+templates = {
+    'junit': os.path.join('reports', 'junit.xml')
+}
+
+#: Allowed report formats
+valid_formats = list(templates.keys())
+
+__all__ = [
+    'valid_formats',
+    'collect_info'
+]
+
+
+def fetch_text(path):
+    if not os.path.exists(path):
+        return 'No such file or directory: {0}'.format(path)
```

Comment:
Looks like the `No such file or directory` would end up in the report in place of the actual test's `stdout` when the latter is missing. When this happens, is there a way to tell the difference between this error message (a kind of *internal error*) and something emitted by the test itself?

Reply:
> Looks like the No such file or directory would end up in the report in place of the actual test's stdout when the latter is missing.

You're right on that. I am a bit undecided what to do, in the sense that I am trying to distinguish `errors` from `failures` based on the exception type. As we have the real `stdout` and `stderr` from the build process merged and dumped into `build_log_path` I was thinking to:

1. map `build_log_path` to `<stdout/>`
2. map the exception stackstrace to `<stderr/>`  - still to be implemented
3. map a fixed message (like I do right now) or maybe `str(exc_val)` to the `message` property.

The end result looks like this:

![junitfailurejenkins](https://user-images.githubusercontent.com/4199709/35170028-0e67d08c-fd5f-11e7-8a57-ecf9b88234ee.png)

on the Jenkins JUnit plugin. Suggestions?

Reply:
The report looks quite clear, imo as long as you have the stack trace any other data becomes (nice to have tho) decor.